src/runtime/long.cpp

```diff
@@ -1393,14 +1424,20 @@ Box* longHash(BoxedLong* self) {
     if (mpz_fits_slong_p(self->n))
         return boxInt(mpz_get_si(self->n));
 
-    // Not sure if this is a good hash function or not;
-    // simple, but only includes top bits:
-    union {
-        uint64_t n;
-        double d;
-    };
-    d = mpz_get_d(self->n);
-    return boxInt(n);
+    // CPython use the absolute value of self mod ULONG_MAX.
+    mpz_t r;
+    mpz_init(r);
+    long remainder = mpz_mmod_ui(r, self->n, ULONG_MAX);
+
+    // if the sign of remainder is positive but self is negtive, it means the remainder overflowed.
+    // So compensate by incrementing. This preserves the value modulo ULONG_MAX.
+    if ((remainder >= 0) && (mpz_sgn(self->n) < 0))
+        remainder++;
+
+    if (remainder == -1)
+        remainder = -2;
+
+    return boxInt(remainder);
```

Comment:
Ok, I think I figured this out :)  The tricky thing is that if `self` is divisible by ULONG_MAX, then CPython's "compute self mod ULONG_MAX" operation will return ULONG_MAX, not 0.  And ULONG_MAX is -1 so it will then get changed to -2.

After some hacking around, this is what I came up with:
```C++
unsigned long remainder = mpz_tdiv_ui(self->n, ULONG_MAX);
if (remainder == 0)
  remainder = -1; // CPython compatibility -- ULONG_MAX mod ULONG_MAX is ULONG_MAX to them.
remainder *= mpz_sgn(self->n);
if (remainder == -1)
  remainder = -2;
return boxInt(remainder);
```

Can we also add this to the test:
```Python
for i in xrange(100):
  for j in xrange(100):
    print i, j, hash((1 << i) - (1 << j))
```